src/machine/aot/aot.x64.c

```diff
@@ -280,67 +317,123 @@ void aot_finalize(AotContext* context)
 int aot_link(AotContext* context, size_t *szp)
 {
   dasm_State** Dst = &context->d;
+
   /*
-   * Check memory write permissions. Note this pseudo function does not use
-   * C's standard calling convention, since the AOT code here has its own
-   * register allocations for maximum performance. Required arguments to this
-   * pseudo function include:
-   *
-   * rax: the memory address to check for permissions
-   * rdx: length of memory to write
-   *
-   * The return value is kept in rdx, 0 means success, while non-zero values
-   * mean permission check fails.
-   *
-   * Note the free register rcx might also be modified in this pseudo function.
-   */
-  |->check_write:
+  * Initialize the memory and perform permission checks. Note this pseudo
+  * function does not use C's standard calling convention, since the AOT
+  * code here has its own register allocations for maximum performance.
+  * Required arguments to this pseudo function include:
+  *
+  * rax: the memory address to check for permissions
+  * rdx: length of memory to write
+  * rcx:
+  *   non-zero: do permission check
+  *   zero: don't do permission check
+  *
+  * The return value is kept in rdx, 0 means success, while non-zero values
+  * mean permission check fails.
+  */
+  |->access_control:
   | push rsi
-  | mov rsi, rdx
+  | push r8
+  | push r9
+  /*
+   * Test if the frame needs to be initialized
+   */
+  | mov rsi, rax
+  | shr rsi, CKB_VM_ASM_MEMORY_FRAME_SHIFTS
+  | cmp rsi, CKB_VM_ASM_MEMORY_FRAMES
+  | jae >2
+  | lea r9, machine->frames
+  | movzx r8d, byte [r9+rsi]
+  | cmp r8d, 0
+  | jne >4
+  | mov byte [r9+rsi], 1
+  | push rax
+  | mov rax, rsi
```

Comment:
Just one minor suggestion here: if `zeroed_memory` takes address from rsi instead of rax, we might not need to save, update and restore rax here, correct?

Reply:
Afraid not, the syscall number needs to be stored in rax.

Reply:
What is syscall number here?

Reply:
https://github.com/nervosnetwork/ckb-vm/pull/107/files/822591565524b1a036cfdbc226618b418ff9c74f#diff-429dcb115a620eba83a6f6512807bfffR433

The address of memset is stored in rax, and rsi may be used as a parameter. **Not a syscall number, I described this code incorrectly**.

Reply:
Wait a minute, I will try to change it, you may be right

Reply:
What I'm saying, is that you can just use rsi as a parameter to `zeroed_memory` instead of rax. Internally in `zeroed_memory` is a different story. This way, you don't need to push rax, mov rax, rsi and then pop rax surrounding zeroed_memory. The address, that will be used by `zeroed_memory`, is already in rsi, and we can rely on this fact.

Reply:
I changed it, the code is more clear than before.